native/server/src/sqlite_brongnal.rs

```diff
@@ -0,0 +1,350 @@
+                (),
+            )
+            .context("Creating pre_key table failed.")?;
+        connection
+            .execute(
+                "CREATE TABLE IF NOT EXISTS message (
+             message BLOB PRIMARY KEY,
+             user_identity STRING NOT NULL,
+             creation_time integer NOT NULL,
+             FOREIGN KEY(user_identity) REFERENCES user(identity)
+         )",
+                (),
+            )
+            .context("Creating message table failed.")?;
+
+        Ok(SqliteStorage(Arc::new(Mutex::new(connection))))
+    }
+}
+
+impl Storage for SqliteStorage {
+    fn add_user(
+        &self,
+        identity: String,
+        identity_key: VerifyingKey,
+        signed_pre_key: proto::service::SignedPreKey,
+    ) -> tonic::Result<()> {
+        eprintln!("Adding user: {identity}");
+
+        let _ = self.connection()?.execute(
+            "INSERT INTO user (identity, key, current_pre_key, creation_time) VALUES (?1, ?2, ?3, ?4)",
+            (
+                identity, identity_key.to_bytes(), signed_pre_key.encode_to_vec(),
+                SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs(),
+            ),
+        ).context("failed to insert key.");
+        Ok(())
+    }
+
+    fn clear_one_time_keys(&self, identity: &str) -> tonic::Result<()> {
+        eprintln!("Clearing one time keys for user: {identity}");
+        let _ = self
+            .connection()?
+            .execute(
+                "DELETE FROM pre_key WHERE user_identity = ?1",
+                params![identity.to_owned()],
+            )
+            .map_err(|e| {
+                Status::internal(format!("failed to clear one time keys for {identity}: {e}"))
+            })?;
+        Ok(())
+    }
+
+    fn update_pre_key(
+        &self,
+        identity: &str,
+        signed_pre_key: proto::service::SignedPreKey,
+    ) -> tonic::Result<()> {
+        eprintln!("Updating pre key for user: {identity}");
+
+        let _: String = self
+            .connection()?
+            .query_row(
+                "UPDATE user SET current_pre_key = ?2 WHERE identity = ?1 RETURNING identity",
+                params![identity, signed_pre_key.encode_to_vec()],
+                |row| Ok(row.get(0)?),
+            )
+            .map_err(|_| Status::not_found("user not found"))?;
+        Ok(())
+    }
+
+    fn add_one_time_keys(
+        &self,
+        identity: &str,
+        pre_keys: Vec<X25519PublicKey>,
+    ) -> tonic::Result<()> {
+        eprintln!(
+            "Adding {} one time keys for user: {identity}",
+            pre_keys.len()
+        );
+
+        let connection = self.connection()?;
+        let mut stmt = connection
+            .prepare("INSERT INTO pre_key (user_identity, key, creation_time) VALUES (?1, ?2, ?3)")
+            .unwrap();
+        for pre_key in pre_keys {
+            stmt.execute((
+                identity,
+                pre_key.to_bytes(),
+                SystemTime::now()
```

Comment:
I don't personally love epoch time, and prefer ISO-8601 for human readability. What are your thoughts?

Reply:
Storing human readable timestamps is silly. It makes the most sense to store epoch time and then convert when it needs to be viewed by a human.